warp/handlers/signature_request_p2p.go

```diff
@@ -0,0 +1,146 @@
+// (c) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package handlers
+
+import (
+	"context"
+	"time"
+
+	"github.com/ava-labs/avalanchego/codec"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p"
+	"github.com/ava-labs/avalanchego/proto/pb/sdk"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/crypto/bls"
+	avalancheWarp "github.com/ava-labs/avalanchego/vms/platformvm/warp"
+	"github.com/ava-labs/subnet-evm/warp"
+	"google.golang.org/protobuf/proto"
+)
+
+var _ p2p.Handler = (*SignatureRequestHandlerP2P)(nil)
+
+const (
+	ErrFailedToParse = iota
+	ErrFailedToGetSig
+	ErrFailedToMarshal
+)
+
+// SignatureRequestHandlerP2P serves warp signature requests using the p2p
+// framework from avalanchego. It is a peer.RequestHandler for
+// message.MessageSignatureRequest.
+type SignatureRequestHandlerP2P struct {
+	backend warp.Backend
+	codec   codec.Manager
+	stats   *handlerStats
+}
+
+func NewSignatureRequestHandlerP2P(backend warp.Backend, codec codec.Manager) *SignatureRequestHandlerP2P {
+	return &SignatureRequestHandlerP2P{
+		backend: backend,
+		codec:   codec,
+		stats:   newStats(),
+	}
+}
+
+func (s *SignatureRequestHandlerP2P) AppRequest(
+	ctx context.Context,
+	nodeID ids.NodeID,
+	deadline time.Time,
+	requestBytes []byte,
+) ([]byte, *common.AppError) {
+	// Per ACP-118, the requestBytes are the serialized form of
+	// sdk.SignatureRequest.
+	req := new(sdk.SignatureRequest)
+	if err := proto.Unmarshal(requestBytes, req); err != nil {
+		return nil, &common.AppError{
+			Code:    ErrFailedToParse,
+			Message: "failed to unmarshal request: " + err.Error(),
+		}
+	}
+
+	var (
+		sig [bls.SignatureLen]byte
+		err error
+	)
+	if len(req.Message) == ids.IDLen {
```

Comment:
The `req.Message` to be signed here will be the raw bytes of the unsigned Warp message, so it will need to be parsed as an `UnsignedMessage` [here](https://github.com/ava-labs/avalanchego/blob/master/vms/platformvm/warp/unsigned_message.go#L39).

Then the `Payload` of the message can be determined to either be an `AddressedCall` or `Hash` based on `typeID`s defined [here](https://github.com/ava-labs/avalanchego/tree/master/vms/platformvm/warp/payload).

Reply:
does that mean we have first parse it for `UnsignedMessage` and then do additinal parsing/handling for `AddressedCall` vs `Hash`?

Reply:
i.e:
```
unsignedWarpMsg := new(UnsignedMessage)
codec.Unmarshal(req.Message, unsignedWarpMsg)

blockhashTtype BlockHashType | AddressedCallType
codec.Unmarshal(blockHashType, payload)
blockHashType.Handle()
```

Reply:
Yeah I believe so. I think we can use the [`warp.ParseUnsignedMessage`](https://github.com/ava-labs/avalanchego/blob/master/vms/platformvm/warp/unsigned_message.go#L39), [`payload.ParseAddressedCall`](https://github.com/ava-labs/avalanchego/blob/master/vms/platformvm/warp/payload/addressed_call.go#L33), and [`payload.ParseHash`](https://github.com/ava-labs/avalanchego/blob/master/vms/platformvm/warp/payload/hash.go#L29) helpers directly.